### PR TITLE
Bug: Domain only validation doesn't work for custom schemes

### LINKED ISSUE #200: Bug: Domain only validation doesn't work for custom schemes
If an application is using a custom scheme (such as for a desktop or mobile application), then just validation of the login domain is not enough.

The Library should instead, validate the URI using this set of rules:

- If the passed in value has a scheme, the scheme must match
- If the passed in value has a path, the path must match, although NOT the query values
- The domain must always match BOTH the URI and the domain values
- Should NOT validate ports

Should Work Examples:

- `https://www.example.com/login` Validate `https`, `www.example.com` and `/login`
- `example://login` Validate `example`, `login`, and Ignore path
- `www.example.com/login` Validate `www.example.com` and `/login` and Ignore scheme
- `localhost:3030/login/path` Validate `localhost` domain and `/login/path` path. Ignore the port

### PR DESCRIPTION
**Pull request by mattheworris**
# Problem

Closes #200

### CODE REVIEW

File: libraries/js/src/mocks/payloads.ts [hunk 1]
```diff
@@ -12,8 +12,8 @@ import {
   serializeItemActionsPayloadHex,
 } from '../util.js';
 
-function generateLoginMessage(account: string, issued: Date, expires: Date, domain: string) {
-  return `${domain} wants you to sign in with your Frequency account:\n${account}\n\n\n\nURI: https://testnet.frequencyaccess.com/signin/confirm\nNonce: N6rLwqyz34oUxJEXJ\nIssued At: ${issued.toISOString()}\nExpiration Time: ${expires.toISOString()}`;
+function generateLoginMessage(account: string, issued: Date, expires: Date, uri: URL, domain: string) {
+  return `${domain} wants you to sign in with your Frequency account:\n${account}\n\n\n\nURI: ${uri}\nNonce: N6rLwqyz34oUxJEXJ\nIssued At: ${issued.toISOString()}\nExpiration Time: ${expires.toISOString()}`;
```
Comment (wilwade) [inline libraries/js/src/mocks/payloads.ts:16]:
```suggestion
function generateLoginMessage(account: string, issued: Date, expires: Date, uri: URL) {
  return `${uri.hostname} wants you to sign in with your Frequency account:\n${account}\n\n\n\nURI: ${uri}\nNonce: N6rLwqyz34oUxJEXJ\nIssued At: ${issued.toISOString()}\nExpiration Time: ${expires.toISOString()}`;
```

File: libraries/js/src/mocks/payloads.ts [hunk 2]
```diff
@@ -24,6 +24,7 @@ const loginMessageGood = (domain: string) =>
     ExampleUserKey.public,
     new Date(now - 24 * 60 * 60 * 1000),
     new Date(now + 24 * 60 * 60 * 1000),
+    new URL('https://testnet.frequencyaccess.com/signin/confirm'),
     domain
```
Comment (wilwade) [inline libraries/js/src/mocks/payloads.ts:28]:
```suggestion
    new URL('https://your-app.com/signin/callback')
```

File: libraries/js/src/payloads.ts [hunk 3]
```diff
@@ -92,6 +114,72 @@ function expect(test: boolean, errorMessage: string) {
...
+  /* eslint-disable prefer-const */
+  let msgParsedScheme: string | null;
+  let expectedParsedScheme: string | null;
+  let msgParsedDomain: string | undefined;
+  let expectedParsedDomain: string | undefined;
+  let msgParsedPath: string | undefined;
+  let expectedParsedPath: string | undefined;
+
+  // Parse out the optional scheme from the domain
+  // SIWF_V2_DOMAIN may be configured with or without a scheme that accommodates app-specific needs.
```
Comment (wilwade) [inline libraries/js/src/payloads.ts:140]: `SIWF_V2_DOMAIN` is not known by this library.

File: libraries/js/src/payloads.ts [hunk 4]
```diff
@@ -92,6 +114,72 @@ function expect(test: boolean, errorMessage: string) {
...
+ * 1. If the domains have schemes (e.g., `https://`), the schemes must match.
+ * 2. The paths within the domains must match.
+ * 3. The domains themselves must match.
+ *
+ * @param msgDomain - The domain from the message to be validated. This can include an optional scheme and path.
+ * @param expectedDomain - The expected domain to validate against. This can include an optional scheme and path.
+ *
+ * @throws Will throw an error if the schemes, paths, or domains do not match.
+ */
+function validateDomain(msgDomain: string, expectedDomain: string) {
```
Comment (wilwade) [inline libraries/js/src/payloads.ts:130]:
```suggestion
function validateDomainAndUri(msgUri: string, expectedUri: string) {
```

**Review by wilwade — CHANGES_REQUESTED** [review]
Few fixes and questions, but I think the core validation logic is correct.

File: libraries/js/src/payloads.ts [hunk 5]
```diff
@@ -109,10 +197,10 @@ function validateLoginPayload(
 
   // Validate the message contents
   const msg = parseMessage(payload.payload.message);
-  expect(
-    msg.domain === loginMsgDomain,
-    `Message does not match expected domain. Message: ${msg.domain} Expected: ${loginMsgDomain}`
-  );
+
+  // Validate the domain
+  validateDomain(msg.domain, loginMsgDomain);
```
[the remark this answers is not in the document]
Comment (mattheworris) [reply libraries/js/src/payloads.ts:202]: Yes, my tests are off, working on updating now.

File: libraries/js/src/mocks/payloads.ts [hunk 6]
```diff
@@ -12,38 +12,52 @@ import {
...
+    message: loginMessageUrl(url),
+  },
+});
+
+export const ExamplePayloadLoginGood = (): SiwfResponsePayloadLogin => ({
   signature: {
     algo: 'SR25519',
     encoding: 'base16',
-    encodedValue: u8aToHex(ExampleUserKey.keyPair().sign(loginMessageGood(domain))),
+    encodedValue: u8aToHex(ExampleUserKey.keyPair().sign(loginMessageGood())),
```
Comment (mattheworris) [inline libraries/js/src/mocks/payloads.ts:56]: domain will always match the URI in the test cases

File: libraries/js/src/payloads.ts [hunk 7]
```diff
@@ -25,6 +25,19 @@ interface SiwxMessage {
   uri: string;
 }
 
+/**
```
Comment (mattheworris) [inline libraries/js/src/payloads.ts:28]: Added: helpful for hover over in IDE

File: libraries/js/src/response.test.ts [hunk 8]
```diff
@@ -19,7 +19,9 @@ describe('getLoginResult', () => {
       text: () => Promise.resolve('MOCK'),
     } as any);
 
-    await expect(getLoginResult('fakeAuthCode', { loginMsgDomain: 'localhost' })).to.resolves.toMatchObject(example);
+    await expect(
+      getLoginResult('fakeAuthCode', { loginMsgUri: 'testnet.frequencyaccess.com' })
```
Comment (mattheworris) [inline libraries/js/src/response.test.ts:23]: `localhost` changed in `ExampleLogin()`

**Review by JoeCap08055 — COMMENTED** [review]
Hmm, I've lost quite a bit of context since SIWFv1, but I just want to point out that `domain` and `uri` are two separate and unrelated things in the "Sign in with Ethereum" (Sign in With X) spec.

`domain` is the domain of the app requesting a signin (ie, mewe.com)
`uri` is the URI of the signing app (ie, frequencyaccess.com)

We should make sure that our use of these fields is as the standard expects; at this point I'm a little unclear if it does.
(Realizing that this may work fine for SIWFv2, but may break when we re-integrate standalone wallets and/or the Talisman SIWS package)

**Review by JoeCap08055 — APPROVED** [review]
My comments/concerns addressed in an external thread; looks good!

File: docs/src/Actions/Response.md [hunk 9]
```diff
@@ -107,7 +108,10 @@ Chain ID: frequency:{{chainReference}}
 Issued At: {{issued-at}}
 ```
 
+Inside the message, `{{domain}}` is the domain of the application requesting the sign-in. `{{domain}}` should match the domain contained in the `URI` field.
```
Comment (wilwade) [inline docs/src/Actions/Response.md:111]: ❤️

Comment (mattheworris) [inline docs/src/Actions/Response.md:111]: Domain clarification

**Review by wilwade — APPROVED** [review]
💯 Thanks!

**Pull request merged**